---
title: "Puntseq WGS visualizations"
output:
  html_document:
    df_print: paged
---

```{r}
library(data.table)
library(tidyr)
library(ggplot2)
library(dplyr)
library(RColorBrewer)
library(scales)
library(stringr)
library(reticulate)
library(Rcpp)
library(forcats)
library(patchwork)
library(xtable)
```
# Quality control 
TODO I do not know how to fix the error on the server; however the plot is anyways complete 

```{r}
require("reticulate")
#only activate once 
#conda_create("pickle", python_version = "3.9")
#use_condaenv("pickle")
#conda_install("pickle", "pandas")

#source_python("workflow/scripts/pickle_reader.py")
#fastq_new_dt <- data.table(read_pickle_file("qc_plots/shotgun_new_NanoPlot-data.pickle"))
#fastq_fahad_dt <- data.table(read_pickle_file("qc_plots/shotgun_fahad_NanoPlot-data.pickle"))
#tables <- list(fastq_new = fastq_new_dt, fastq_fahad=fastq_fahad_dt)
#runs_dt <- rbindlist(tables, idcol = "run")
#runs_dt$run <- str_replace(runs_dt$run, 'fastq_new', "Guppy v6.3.7+532d626")
#runs_dt$run <- str_replace(runs_dt$run, 'fastq_fahad', "Guppy v4.0.14")

#ggplot(data=runs_dt, aes(lengths, quals)) + geom_bin2d() + scale_fill_continuous(type = "run") + geom_vline(xintercept = 1500) + scale_x_log10(labels=comma) + theme_bw() + facet_wrap(~run) + xlab("Read lengths") + ylab("Average read quality")
``` 

# MAGpy shotgun
## Prior knowledge 
```{r}
PATHOGENIC_GENERA <- c("Arcobacter", "Aeromonas", "Pseudomonas", 
                       "Legionella", "Escherichia", "Bacillus", 
                       "Serratia", "Klebsiella", "ENterobacter", 
                       "Yersinia", "Acinetobacter", "Coxiella",
                       "Salmonella", "Streptococcus", "Enterococcus",
                       "Clostridium", "Citrobacter", "Stenotrophomonas",
                       "Listeria", "Leptospira")
```
## CheckM
Tidy a data table. 
```{r}
checkm_dt <- fread("MAGpy/checkm_plus.txt")
# bin_id contains the sample identifier and bin number, which we need in a separate column 
checkm_dt <- checkm_dt %>% separate(col = "bin_id", into = c("shotgun", "sampletmp", "binning_tool", "bin_id")) %>% unite(col=sample, 
                                                                                                             shotgun, 
                                                                                                             sampletmp, sep = "_") 
checkm_dt$sample <- str_replace(checkm_dt$sample, 'shotgun_new', "Guppy v6.3.7")
checkm_dt$sample <- str_replace(checkm_dt$sample, 'shotgun_fahad', "Guppy v4.0.14")
```
General distribution of completeness and contamination across all bins
```{r}
completeness <- ggplot(data = checkm_dt, aes(x=sample, y=completeness)) + 
  geom_violin() + 
  geom_point() + 
  facet_wrap(~binning_tool) + 
  guides(x =  guide_axis(angle = 45)) +
  theme(axis.title.x = element_blank()) 

contamination_reduced <- ggplot(data = checkm_dt[contamination <= 100,], aes(x=sample, y=contamination)) + 
  geom_violin() + 
  geom_point() + 
  facet_wrap(~binning_tool) + 
  guides(x = guide_axis(angle = 45)) + 
  theme(axis.title.x = element_blank()) 

completeness / contamination_reduced 
```
Number of bins
```{r}
cast_completeness <- function(completeness){
  if (completeness > 90) {
    return("compl90")
  }
  else if (completeness > 80) {
    return("compl80")
  }
  else if (completeness > 70) {
    return("compl70")
  }
  else if (completeness > 60) {
    return("compl60")
  }
  else {
    return("incomplete")
  }
}
bin_completeness <- checkm_dt %>% .[, stack := sapply(completeness, cast_completeness)] %>% .[, .(num_bins = .N), by=.(stack, binning_tool, sample)] %>% .[order(stack),]

ggplot(data = bin_completeness, aes(x=binning_tool, y=num_bins, fill=stack)) + 
  geom_bar(stat = "identity", 
           position="stack") + 
  ylab("number of bins") + 
  scale_fill_brewer(palette="Paired") +
  theme_minimal() + 
  guides(fill=guide_legend(title = "binning tool")) + #change legend title 
  theme(axis.title.x = element_blank()) +
  facet_grid(~ sample)
``` 

## Sourmash 
Tidy a data table
```{r}
sourmash_dt <- fread("MAGpy/sourmash_report.csv")
sourmash_dt <- sourmash_dt %>% separate(ID, into = c(NA, "shotgun", "sample", "binning_tool", "bin_id", NA)) %>% unite(col=sample, shotgun, sample, sep = "_")

sourmash_dt$sample <- str_replace(sourmash_dt$sample, 'shotgun_new', "Guppy v6.3.7")
sourmash_dt$sample <- str_replace(sourmash_dt$sample, 'shotgun_fahad', "Guppy v4.0.14")
```
Bins summary plot 
```{r}
bins_summary_plot <- ggplot(data = sourmash_dt[genus != ""], aes(x = fct_infreq(genus), fill = binning_tool)) + geom_bar(stat="count", position=position_dodge()) + ylab("Number of bins") + facet_grid(~sample) 
bins_summary_plot + xlab("") + coord_flip() + ggtitle("Predicted genus by sourmash")
```
## DIAMOND
Tidy the data table 
```{r}
diamond_dt <- fread("MAGpy/diamond_bin_report_plus.tsv")

# genus column occurs twice 
colnames(diamond_dt)[5] = "estimated_genus"
diamond_dt <- diamond_dt %>% separate(., name, into = c("shotgun", "sample", "binning_tool", "bin_id")) %>% unite(col=sample, shotgun, sample, sep = "_")
diamond_dt$sample <- str_replace(diamond_dt$sample , 'shotgun_new', "Guppy v6.3.7")
diamond_dt$sample <- str_replace(diamond_dt$sample , 'shotgun_fahad', "Guppy v4.0.14")
diamond_dt[,.N, by=.(family, sample)]
```
Bins summary plot 
```{r}
ggplot(data = diamond_dt[genus != ""], aes(x = fct_infreq(genus), fill = binning_tool)) +
  geom_bar(stat="count") + 
  ylab("Number of bins") + 
  facet_grid(~sample) + 
  xlab("") + 
  theme(axis.title.y=element_blank(),
        axis.text.y=element_blank(),
        axis.ticks.y=element_blank()) +
  coord_flip() + 
  ggtitle("Predicted genus (shotgun)")

```
## Shotgun assembly statistics 
[TODO] Size distrbution [bp] of the bins per assembly tool, number of reads [#] per bin (e.g. boxplots)
[TODO] Relative amount of lost reads
[TODO] Replot old and new side-by-side 

Old data 
```{r}
contigs_dt <- fread('results/current/vamb/shotgun_fahad/bins/contig_lengths.csv')
bins <- fread('results/current/vamb/shotgun_fahad/bins/clusters.tsv')
colnames(bins) <- c("bin", "contignames")
colnames(contigs_dt) <- c("index", "contignames", "lengths")

ggplot(data=contigs_dt[lengths>50000], aes(lengths)) + geom_histogram() + scale_x_log10(labels=comma) + scale_y_log10()

contig_bin_dt <- merge.data.table(contigs_dt, bins, by = "contignames" )
contig_bin_dt
contig_bin_dt$sample <- 'shotgun_fahad'
contig_bin_dt[,.N, by=bin] %>% .[N != 1]
bin_lengths <- contig_bin_dt[,.(bin_lengths = sum(lengths), sample), by=.(bin)]
ggplot(data = bin_lengths, aes(bin_lengths)) + geom_histogram() + scale_x_log10(labels=comma) + scale_y_log10()
#ggplot(data = bin_lengths, aes(sample = bin_lengths)) + geom_qq(distribution =  stats::qunif)  + scale_y_log10() + geom_abline(slope=1, intercept=0) + facet_wrap(~sample)
ggplot(data = bin_lengths, aes(x=sample, y=bin_lengths)) + geom_violin() + scale_y_log10(labels=comma)
```
New data 
```{r}
contigs_dt <- fread('results/current/vamb/shotgun_new/bins/contig_lengths.csv')
bins <- fread('results/current/vamb/shotgun_new/bins/clusters.tsv')
colnames(bins) <- c("bin", "contignames")
colnames(contigs_dt) <- c("index", "contignames", "lengths")

ggplot(data=contigs_dt[lengths>50000], aes(lengths)) + 
  geom_histogram() + 
  scale_x_log10(labels=comma) + 
  scale_y_log10() +
  xlab("bp") +
  ylab("bins")

contig_bin_dt <- merge.data.table(contigs_dt, bins, by = "contignames" )
contig_bin_dt
contig_bin_dt$sample <- 'shotgun_new'
contig_bin_dt[,.N, by=bin] %>% .[N != 1]
bin_lengths <- contig_bin_dt[,.(bin_lengths = sum(lengths), sample), by=.(bin)]

ggplot(data = bin_lengths, aes(bin_lengths)) + 
  geom_histogram() + 
  scale_x_log10(labels=comma) + 
  scale_y_log10() + 
  xlab("bp")

#ggplot(data = bin_lengths, aes(sample = bin_lengths)) + geom_qq(distribution =  stats::qunif)  + scale_y_log10() + geom_abline(slope=1, intercept=0) + facet_wrap(~sample)

ggplot(data = bin_lengths, aes(x=sample, y=bin_lengths)) + 
  geom_violin() + 
  scale_y_log10(labels=comma) + 
  xlab("bp")
```
# 16S EMU
Tidy a data table.
```{r}
# concatenate all 16S runs into one table 
files <- list.files(path = 'emu/results',
                    full.names = TRUE,
                    pattern= "*_rel-abundance.tsv") 
# name the list elements by the filenames 
names(files) <- basename(files)
# read all files at once into a list of data.tables
tables <- lapply(files, fread)
# bind all tables into one using rbindlist, 
# keeping the list names (the filenames) as an id column. 
amplicon_dt <- rbindlist(tables, idcol = 'filepath')
amplicon_dt <- amplicon_dt %>% separate(filepath, c("barcode", "sample", NA))
amplicon_dt$sample <- str_replace(amplicon_dt$sample, 'new', "Guppy v6.3.7")
amplicon_dt$sample <- str_replace(amplicon_dt$sample, 'porechop', "Guppy v4.0.14")
# relative abundance in percent
amplicon_dt[, abundance := abundance*100]
```
Not possible to show all genera because too many, but we can separate the visualization into several parts. 
## Most abundant genera
```{r}
ggplot(data = amplicon_dt[abundance > 1, ], aes(x=reorder(genus, desc(abundance)), y=abundance, fill=sample)) + 
  geom_bar(stat="sum", position="dodge") + 
  xlab("") + 
  ylab("abundance %") +
  coord_flip() + 
  ggtitle("Most abundand genera > 1% (16S)") + 
  facet_wrap(~barcode) +
  guides(size = FALSE) # remove n=1
```
## Pathogenic genera
```{r}
ggplot(data = amplicon_dt[genus %in% PATHOGENIC_GENERA,], 
       aes(x=reorder(genus, desc(abundance)), 
           y=abundance, 
           fill=sample)) + 
  geom_bar(stat="sum", position="dodge", width = 0.7)  +
  xlab("") + 
  ylab("abundance %") +
  coord_flip() +
  scale_y_continuous(guide = guide_axis(angle = 45)) +
  ggtitle("Potentially pathogenic genera (16S)") + 
  facet_wrap(~barcode) + 
  guides(size = FALSE)
```

## Leptospira 
```{r}
ggplot(data = amplicon_dt[genus == "Leptospira", ],  
       aes(x=reorder(species, desc(abundance)), 
           y=abundance, 
           fill=sample)) + 
  geom_bar(stat="sum", position="dodge", width = 0.7)  +
  xlab("") + 
  ylab("abundance %") +
  coord_flip() +
  scale_y_continuous(guide = guide_axis(angle = 45)) +
  ggtitle("Leptospira spp. (16S)") + 
  facet_wrap(~barcode) + 
  guides(size = FALSE)
```
## Leptospira 
```{r}
ggplot(data = amplicon_dt[genus == "Arcobacter", ],  
       aes(x=reorder(species, desc(abundance)), 
           y=abundance, 
           fill=sample)) + 
  geom_bar(stat="sum", position="dodge", width = 0.7)  +
  xlab("") + 
  ylab("abundance %") +
  coord_flip() +
  scale_y_continuous(guide = guide_axis(angle = 45)) +
  ggtitle("Arcobacter spp. (16S)") + 
  facet_wrap(~barcode) + 
  guides(size = FALSE)
```
## Overlaps 
```{r}
set1 <- c(unique(diamond_dt[genus != "", genus]), unique(sourmash_dt[genus != "", genus]))
set2 <- unique(amplicon_dt[genus != "", genus])

lst <- list(WGS_sourmash_diamond=set1,
            amplicon_emu=set2)

# load ggVennDiagram Package
library("ggVennDiagram")
  
# use list as an input
  
# creating Venn diagram and displaying 
# all sets
ggVennDiagram(lst) 
```

## 16S vs WGS 
Load the coverm abundance estimations of WGS 
```{r}
# concatenate fahad and new data to one df 
files <- list.files(path = 'results/current/coverm/',
                    full.names = TRUE,
                    pattern= "*.txt") 
# name the list elements by the filenames 
names(files) <- basename(files)
# read all files at once into a list of data.tables
tables <- lapply(files, fread)
# bind all tables into one using rbindlist, 
# keeping the list names (the filenames) as an id column. 
coverm_dt <- rbindlist(tables, idcol = 'filepath')

coverm_dt$filepath <- str_replace(coverm_dt$filepath, 'shotgun_new_summary.txt', "Guppy v6.3.7")
coverm_dt$filepath <- str_replace(coverm_dt$filepath, 'shotgun_fahad_summary.txt', "Guppy v4.0.14")

setnames(coverm_dt, c("sample", "genome", "abundance"))
coverm_dt
```
Load the GTDB taxonomy 
```{r}
# concatenate fahad and new data to one df 
files <- c('results/current/gtdbtk/shotgun_new/shotgun_new.bac120.summary.tsv', 'results/current/gtdbtk/shotgun_fahad/shotgun_fahad.bac120.summary.tsv') 
# name the list elements by the filenames 
names(files) <- basename(files)
# read all files at once into a list of data.tables
tables <- lapply(files, fread)
# bind all tables into one using rbindlist, 
# keeping the list names (the filenames) as an id column. 
taxonomy_dt <- rbindlist(tables, idcol = 'filepath')

taxonomy_dt$filepath <- str_replace(taxonomy_dt$filepath, "shotgun_new.bac120.summary.tsv", "Guppy v6.3.7")
taxonomy_dt$filepath <- str_replace(taxonomy_dt$filepath, "shotgun_fahad.bac120.summary.tsv", "Guppy v4.0.14")

taxonomy_dt <- taxonomy_dt[, .(filepath, user_genome, classification)] %>% setnames(., c("sample", "genome", "classification")) %>% separate(., col = classification, sep = ';',
                                                                                                                                    into = c("superkingdom", "phylum",
                                                                                                                                             "class", "order", 
                                                                                                                                             "family", "genus", 
                                                                                                                                             "species"))
taxonomy_dt$superkingdom <- str_replace(taxonomy_dt$superkingdom, "d__", "")
taxonomy_dt$phylum <- str_replace(taxonomy_dt$phylum, "p__", "")
taxonomy_dt$class <- str_replace(taxonomy_dt$class, "c__", "")
taxonomy_dt$order <- str_replace(taxonomy_dt$order, "o__", "")
taxonomy_dt$family <- str_replace(taxonomy_dt$family, "f__", "")
taxonomy_dt$genus <- str_replace(taxonomy_dt$genus, "g__", "")
taxonomy_dt$species <- str_replace(taxonomy_dt$species, "s__", "")
taxonomy_dt[taxonomy_dt == ""] <- NA
```

Compose clean shotgun data

```{r}
shotgun_dt <- merge.data.table(taxonomy_dt, coverm_dt, by = c("sample", "genome"))
shotgun_dt[order(abundance, decreasing=T),]
```
BugSeq taxonomy profiling from read-based shotgun data. 
```{r}
bugseq_dt <- fread('results/current/bugseq/metagenomic_classification/read-based/shotgun_new.filtered-reads.kreport')
colnames(bugseq_dt) <- c('abundance', 'read_count_taxon_and_below', 'read_count_taxon', 'rank', 'ncbi_id', 'name')

bugseq_dt[(rank == 'P') | (name == 'unclassified'), sum(abundance)] # 99.41% total abundance because some of the low abundant taxa were rounded to 0
#recalculate abundance from read counts 

N_reads <- bugseq_dt[(name == 'unclassified') | (name == 'root'), sum(read_count_taxon_and_below)]

bugseq_phylum_abun_dt <- bugseq_dt[(rank == 'P') | (name == 'unclassified'), .(phylum = name, 
                                                      shotgun_read = read_count_taxon_and_below*100/N_reads)]
# add superkingdom annotation
kingdom_phylum <- bugseq_dt[rank %in% c('D','P'), name]
indices <- which(kingdom_phylum %in% c('Bacteria', "Eukaryota", "Archaea", "Viruses"))
kingdom_lst <- list(Bacteria = kingdom_phylum[(indices[1]+1): (indices[2]-1)],
                    Eykaryota = kingdom_phylum[(indices[2]+1):(indices[3]-1)],
                    Archaea = kingdom_phylum[(indices[3]+1):(indices[4]-1)],
                    Viruses = kingdom_phylum[(indices[4]+1):length(kingdom_phylum)])


bugseq_phylum_abun_dt[phylum %in% kingdom_lst$Bacteria, superkingdom := 'Bacteria']
bugseq_phylum_abun_dt[phylum %in% kingdom_lst$Eykaryota, superkingdom := 'Eykaryota']
bugseq_phylum_abun_dt[phylum %in% kingdom_lst$Archaea, superkingdom := 'Archaea']
bugseq_phylum_abun_dt[phylum %in% kingdom_lst$Viruses, superkingdom := 'Viruses']

#convert to old NCBI taxonomy 
# check Planctomyce -> should be the same 
bugseq_phylum_abun_dt$phylum <- str_replace(bugseq_phylum_abun_dt$phylum, 'Bacteroidota', 'Bacteroidetes')
bugseq_phylum_abun_dt$phylum <- str_replace(bugseq_phylum_abun_dt$phylum, 'Chloroflexota', 'Chloroflexi')
bugseq_phylum_abun_dt$phylum <- str_replace(bugseq_phylum_abun_dt$phylum, 'Dependentiae', 'Candidatus Dependentiae')
bugseq_phylum_abun_dt$phylum <- str_replace(bugseq_phylum_abun_dt$phylum, 'Nitrospirota', 'Nitrospirae')
bugseq_phylum_abun_dt$phylum <- str_replace(bugseq_phylum_abun_dt$phylum, 'Patescibacteria', 'Candidatus Gracilibacteria')
bugseq_phylum_abun_dt$phylum <- str_replace(bugseq_phylum_abun_dt$phylum, 'Verrucomicrobiota', 'Verrucomicrobia')
bugseq_phylum_abun_dt$phylum <- str_replace(bugseq_phylum_abun_dt$phylum, 'Actinomycetota', 'Actinobacteria')
bugseq_phylum_abun_dt$phylum <- str_replace(bugseq_phylum_abun_dt$phylum, 'Bacillota', 'Firmicutes')
bugseq_phylum_abun_dt$phylum <- str_replace(bugseq_phylum_abun_dt$phylum, "Tenericutes", "Mycoplasmatota")
bugseq_phylum_abun_dt$phylum <- str_replace(bugseq_phylum_abun_dt$phylum, "Pseudomonadota", "Proteobacteria")
bugseq_phylum_abun_dt[phylum == 'unclassified', superkingdom := 'unclassified']

bugseq_phylum_abun_dt
```

Shotgun taxonomy profiling.

```{r}
# extract abundances 
shotgun_new_phylum_abun_dt <- shotgun_dt[sample == "Guppy v6.3.7", c("superkingdom", "phylum", "abundance")] %>% .[,.(sum(abundance),
                                                                                                    sample = "shotgun_assembly"), by=c("phylum", "superkingdom")] %>% select(superkingdom, phylum, sample, V1) %>% setnames(., c("superkingdom", "phylum", "sample", "abundance"))
# add unclassified from coverm
shotgun_new_phylum_abun_dt <- rbindlist(list(shotgun_new_phylum_abun_dt, data.table(phylum ="unclassified", 
                                                                                    superkingdom ="unclassified",
                                                                                    sample = "shotgun_assembly", 
                                                                                    abundance = coverm_dt[sample == 'Guppy v6.3.7' & genome == 'unmapped', abundance])))
# convert to wide format
shotgun_new_phylum_abun_dt <- dcast(shotgun_new_phylum_abun_dt, ... ~ sample, value.var = "abundance")
#replace old NCBI taxonomy 
# [TODO] automatically 
shotgun_new_phylum_abun_dt$phylum <- str_replace(shotgun_new_phylum_abun_dt$phylum, 'Actinobacteriota', 'Actinobacteria')
shotgun_new_phylum_abun_dt$phylum <- str_replace(shotgun_new_phylum_abun_dt$phylum, 'Bacteroidota', 'Bacteroidetes')
shotgun_new_phylum_abun_dt$phylum <- str_replace(shotgun_new_phylum_abun_dt$phylum, 'Chloroflexota', 'Chloroflexi')
shotgun_new_phylum_abun_dt$phylum <- str_replace(shotgun_new_phylum_abun_dt$phylum, 'Dependentiae', 'Candidatus Dependentiae')
shotgun_new_phylum_abun_dt$phylum <- str_replace(shotgun_new_phylum_abun_dt$phylum, 'Nitrospirota', 'Nitrospirae')
shotgun_new_phylum_abun_dt$phylum <- str_replace(shotgun_new_phylum_abun_dt$phylum, 'Patescibacteria', 'Candidatus Gracilibacteria')
shotgun_new_phylum_abun_dt$phylum <- str_replace(shotgun_new_phylum_abun_dt$phylum, 'Verrucomicrobiota', 'Verrucomicrobia')
shotgun_new_phylum_abun_dt
```
Amplicon taxonomy profiling 
```{r}
# extract abundances 
amplicon_new_phylum_abun_dt <- amplicon_dt[sample == "Guppy v6.3.7", c("superkingdom", "phylum", "abundance", "barcode")] %>% .[,sum(abundance), by=c("superkingdom", "phylum", "barcode")] %>% setnames(., c("superkingdom", "phylum", "sample", "abundance"))

# add unclassified 
amplicon_new_phylum_abun_dt <- replace(amplicon_new_phylum_abun_dt, amplicon_new_phylum_abun_dt=='', "unclassified")

# convert to wide format
amplicon_new_phylum_abun_dt <- dcast(amplicon_new_phylum_abun_dt, ... ~ sample, value.var = "abundance")
amplicon_new_phylum_abun_dt <- amplicon_new_phylum_abun_dt[!((BC02 == 0) & (BC03 == 0) & (BC04 ==0)),]
amplicon_new_phylum_abun_dt[phylum == 'unclassified', superkingdom := 'unclassified']
amplicon_new_phylum_abun_dt
```
Merged taxonomy profile from amplicon, shotgun_read and shotgun_assembly 

```{r}

comparison_dt <- merge.data.table(shotgun_new_phylum_abun_dt, 
                 amplicon_new_phylum_abun_dt, 
                 all=TRUE,
                 by=c("superkingdom", "phylum"),
                 allow.cartesian=TRUE)

comparison_dt <- merge.data.table(comparison_dt, 
                 bugseq_phylum_abun_dt, 
                 all=TRUE,
                 by=c("superkingdom", "phylum"),
                 allow.cartesian=TRUE)
comparison_dt[is.na(comparison_dt)] <- 0

comparison_dt

mat <- as.matrix(comparison_dt[, c( "shotgun_assembly",
                                    "shotgun_read",
                                   "BC02",
                                   "BC03",
                                   "BC04")])

rownames(mat) <- comparison_dt$phylum
colnames(mat) <- c( "shotgun_assembly", "shotgun_read", "amplicon.BC02", "amplicon.BC03", "amplicon.BC04")
head(mat)

```
Create row annotation 
```{r}
# create the row annotation data frame
row.ann <- data.frame(superkingdom = comparison_dt[, superkingdom])
rownames(row.ann) <- comparison_dt$phylum 
row.ann
```

Plot the comparison 

```{r}
library(pheatmap) 
library(RColorBrewer)

real_zero_color <- brewer.pal(n=9, name='Pastel1')[9]
paletteLength <- 9

color <- colorRampPalette(c(real_zero_color, brewer.pal(n = 4, name = "Blues")))(paletteLength)
annotation_colors <- brewer.pal(n=7, name='Pastel1')[4:7]


# length(breaks) == length(paletteLength) + 1
# use floor and ceiling to deal with even/odd length pallettelengths

breaks <- c(seq(min(mat), 0.00000001, length.out=ceiling(paletteLength/2) + 1), 
              seq(max(mat)/paletteLength, 100, length.out=floor(paletteLength/2)))

pheatmap(mat, 
         color = color,
         breaks = breaks,
         cluster_rows=TRUE, 
         cluster_cols=TRUE,
         fontsize_row = 8,
         angle_col = 45,
         display_numbers = TRUE,
         annotation_row = row.ann,
         annotation_colors = list(superkingdom = c(Archaea=annotation_colors[1],
                                                  Bacteria=annotation_colors[2],
                                                  Eykaryota=annotation_colors[3],
                                                  Viruses=annotation_colors[4],
                                                  unclassified='white')),
         cellheight = 8,
         filename = 'all_abund_mat.png'
         )
```

Only classified
```{r}
rescale_abundance <- function(abundance) {100*abundance/sum(abundance)}

comparison_low_abun_dt <- comparison_dt[(phylum != 'unclassified'),] %>% .[,.(scaled_shotgun_assembly = rescale_abundance(shotgun_assembly), scaled_shotgun_read = rescale_abundance(shotgun_read),
                                                                              scaled_BC02 = rescale_abundance(BC02),
                                                                              scaled_BC03 = rescale_abundance(BC03),
                                                                              scaled_BC04 = rescale_abundance(BC04),
                                                                              phylum, superkingdom)]
low_abun_mat <- as.matrix(comparison_low_abun_dt[, c( "scaled_shotgun_assembly",
                                                      "scaled_shotgun_read",
                                                      "scaled_BC02",
                                                      "scaled_BC03",
                                                      "scaled_BC04")])
rownames(low_abun_mat) <- comparison_low_abun_dt$phylum
colnames(low_abun_mat) <- c( "shotgun_assembly", "shotgun_read", "amplicon.BC02", "amplicon.BC03", "amplicon.BC04")
head(low_abun_mat)


# create the row annotation data frame
row.ann <- data.frame(superkingdom = comparison_low_abun_dt[, superkingdom])
rownames(row.ann) <- comparison_low_abun_dt$phylum 

pheatmap(low_abun_mat, 
         color = color,
         breaks = breaks, 
         cluster_rows=TRUE, 
         cluster_cols=TRUE,
         fontsize_row = 8,
         angle_col = 45,
         display_numbers = TRUE,
         annotation_row = row.ann,
         annotation_colors = list(superkingdom = c(Archaea=annotation_colors[1],
                                                   Bacteria=annotation_colors[2],
                                                   Eykaryota=annotation_colors[3],
                                                   Viruses=annotation_colors[4])),
         cellheight = 8,
         filename = 'low_abund_mat.png') 

```
## In-detail comparison investigation 
Two assembled phyla lacking in 16S and shotgun read mapping approaches. 
```{r}
target_phyla <- shotgun_dt[phylum %like% 'Dependentiae',] %>% separate(., genome, into = c('binning_tool', 'bin_id','contigs')) %>% .[, .(sample, binning_tool, bin_id = as.numeric(bin_id), superkingdom, phylum)]

checkm_dt[, bin_id := as.numeric(bin_id)]

das_tools_eval <- fread("results/current/das_tool/shotgun_new/shotgun_new_metabat.eval")
das_tools_eval <- das_tools_eval %>% separate(., bin, into = c('binning_tool', 'bin_id')) 
das_tools_eval[, bin_id := as.numeric(bin_id)]
           
target_phyla_info <- merge.data.table(target_phyla, checkm_dt, 
                 by.x = c('sample', 'binning_tool', 'bin_id'),
                 by.y = c('sample', 'binning_tool', 'bin_id'))
target_phyla_info <- merge.data.table(target_phyla_info, das_tools_eval, 
                 by.x = c('binning_tool', 'bin_id'),
                 by.y = c('binning_tool', 'bin_id')) %>% .[,.(binning_tool, bin_id, sample, superkingdom = superkingdom.x, phylum = phylum.x, size, contigs, N50, n_genomes, n_markers, n_marker_sets, completeness, contamination, binScore, uniqueBacSCGs, uniqueArcSCGs   )]
#print(xtable(target_phyla_info), include.rownames = TRUE, include.colnames = TRUE, sanitize.text.function = I)
target_phyla_info
```